src/metabase/upload.clj

```diff
@@ -232,14 +274,26 @@
 (defn- parse-rows*
   "Returns a lazy seq of parsed rows, given a sequence of upload types for each column.
   Replaces empty strings with nil."
-  [col-upload-types rows]
+  [col-types rows]
   (let [settings (upload-parsing/get-settings)
-        parsers  (map #(upload-parsing/upload-type->parser % settings) col-upload-types)]
+        parsers  (map #(upload-parsing/upload-type->parser % settings) col-types)]
     (for [row rows]
       (for [[value parser] (map-with-nils vector row parsers)]
         (when (not (str/blank? value))
           (parser value))))))
 
+(mu/defn ^:private column-types-from-rows :- [:sequential (into [:enum] column-types)]
+  "Returns a sequence of types, given the unparsed rows in the CSV file"
+  [settings column-count rows]
+  (->> rows
+       (map #(row->value-types % settings))
+       (reduce coalesce-types (repeat column-count nil))
+       (map (fn [type]
+              ;; if there's no values in the column, assume it's a string
+              (if (nil? type)
+                ::text
+                (column-type type))))))
```

Comment:
Ehhh IDK feels like putting `nil ::text` as a case in `column-type` would be nice to get cleaner code (just `(map column-type)` here at the price of obscuring the logic somewhat.

Reply:
Maybe, can you explain why? The idea is to keep the the logic for handling the edge case of an empty column separate from the logic in `column-type`, which handles when there is at least one value in a column.

If I put the logic in `column-type`, I'd have to explain in `column-type`'s docstring that a `nil` value for `type` corresponds to an empty column. That seems to couple `column-type`'s implementation to the implementation of `coalesce-types` a bit, and I think that doing it this way avoids that coupling.

Reply:
I looked at it from the point where if data type is unknown (`nil`), then column type should be `text`. But, as I've said, "nicer code at the price of obscuring the logic". :)